### DISCUSSION TITLE
FPS aquired vs fps written for play back

### DISCUSSION THREAD

**Discussion by aaronjnorris** [bonsai-rx/bonsai#1112] [General]
We have had difficulty using general USB cameras for recording videos and then having the length of the recorded video match the time length of the recording because the frame rate set can be set within the video capture node doesn't match the rate actually achieved.  We made the script here for troubleshooting this problem.  I have seen other discussions on the topic but have not seen any solutions that might work more generally. I would like to have the video writer write at the speed the frames are obtained. Any suggestions on how to do this?
We have been using ELP and Logitech cameras to avoid costs and allow for scaling. We do not need particularly high frame rates (30-60 would be enough). The ability to control these cameras via Bonsai varies and the set frame rate is often not attained in the recording.    
I have seen that the interframe interval and be quite variable presumably due to computer delays. Any suggestions on how to control the frame rates of these cameras other than options in the video writer node that appear?

Would also love suggestions on the linked script that adds frame rate, number, and interval to the video!

![image](https://user-images.githubusercontent.com/29385322/199847320-7e0011f6-3da2-44a7-b1f0-16cea08b933b.png)
https://1drv.ms/u/s!AsluozMIo0dagfF3Mx1BAtilJsyLvA?e=KyZRgh

**Comment by bruno-f-cruz** [comment]
Have you tried to adjust the exposure (make sure it is not doing automatic exposure)

**Reply by aaronjnorris** [reply]
yes - helps maintain consistency but not always getting frames at the rate set in Bonsai

**Comment by jbscherrer** [comment]
The framerate of the camera device is set by an internal clock in the camera itself and will be independent of camera-computer communication timing. The interframe interval variation that you see in the script is due to latency in USB communication and CPU timing, or as Bruno has said, potentially automatic exposure on the camera chip.

**Comment by aaronjnorris** [comment]
Thanks for the input and help. Is there a way to get Bonsai to record video for playback at the rate frames arrive? If the rate is slightly less than set in the video writer (ex actual 28fsp vs set in writer at 30 fps) then the videos end up short.

**Comment by bruno-f-cruz** [comment]
You can change the framerate parameter in videowriter to 28.

**Reply by chongtianyifa** [reply]
I have this issue for a very long time. For a 21 minute recording, the length of written video is 20:42 s. When I do the recording using OBS software, the recording length match exactly as the real duration while the framerate is kept at 30 fps. Is this still a issue of internal clock in the camera itself?

**Reply by jbscherrer** [reply]
The frames will always be acquired at the rate of the camera clock, which will be set by a PLL to be as close as possible to the target fps. So 30 fps will in practice be, say, 29.9 fps, and can also drift around with chip temperature.

Every frame captured by the camera will be encoded by Bonsai along with video file metadata that includes the playback framerate. When you play back the file using a video player, your computer will play back the frames in sequence at the framerate specified in the metadata. If you are recording at 29.9 fps and encoding at 30 fps, this will cause a discrepancy in timing. For any actual experiments you should always be recording timestamps for the incoming frames and using a frame sync signal to synchronize the frames to your other data streams.

**Reply by chongtianyifa** [reply]
I understand that the actual rate that a camera captures image will vary. However, no matter how unstable it is, the number of frames saved in a specific period should not be dependent on which software you use if the software can really save all frames. Then I don't understand why I got different frames using  Bonsai vs OBS. For OBS, the number of frames=fps * real recording duration while in bonsai, the number of frames is always less than fps * real recording duration? In another word, my camera is able to capture at 30 fps (showed by OBS) and Bonsai dropped some frames compared with OBS, although buffed is turned on for Bonsai Videowritter.

**Reply by bruno-f-cruz** [reply]
How are you calculating the number of saved frames? What do you mean by "real recording duration"? The duration of the video or the real time? If the latter, how are you making sure you record for the same time in both softwares? Can you use python to load the video in python and check the total number of frames?

**Reply by chongtianyifa** [reply]
To rule out the possibility of reducing processing speed by any nodes, here is my simple workflow:
![image](https://user-images.githubusercontent.com/46367566/201529271-949bb19d-5bf4-43d5-a5dd-c1ff110c0d85.png)
I count the frame number using matlab: v=VideoReader(filename), the number of frames can be found by its property: 
![image](https://user-images.githubusercontent.com/46367566/201529406-a3aaab4f-332b-4486-a50a-96421fa81e01.png)
This result was obtained yesterday by recording for 2 minutes through manual control (staring at stopwatch).

I also tried using a 2-minute timer in bonsai to precisely control the recording duration:
![image](https://user-images.githubusercontent.com/46367566/201529534-3a320147-434b-48f6-9886-cb5fff3d4fb9.png)
Here is the result:
![image](https://user-images.githubusercontent.com/46367566/201529994-b46fdf04-6af8-4915-a025-30915fb3c326.png)

Both methods generated a video with a length of ~118s.

I also tested using OBS to record for 2 minute at the same context and it give me a 2-minute video with correct frame numbers:
![image](https://user-images.githubusercontent.com/46367566/201530547-9b123ca4-4021-4a5c-8ebf-a2f4b0064c20.png)

[bonsai.zip](https://github.com/bonsai-rx/bonsai/files/9997576/bonsai.zip)

**Reply by bruno-f-cruz** [reply]
Looking at your workflow, it makes sense the results you are seeing. As [jbscherrer](https://github.com/jbscherrer) mentioned, your camera is likely acquiring frames at a slightly higher speed than 30fps. Now, why is it that OBS gives you exactly 2min? I think it has to do with the way the two software determine the duration of the output video. Bonsai doesn't care about the duration of the .avi file, instead, it uses the computer clock to start and terminate the process. If you get 120 or 110 frames it wont matter for Bonsai, since you will always save the frames that are determined by the timer. An alternative way for defining the duration of the video is to establish an output fps of the .avi (say 30fps) and the wanted duration (say 1min). In this case, I can make the choice of recording EXACTLY that number of frames (i.e. 30fps * 60seconds) that will produce this duration after encoding the video. This will give me a video of exactly 1minute for instance. While the duration of the video will be 1min, it might not correspond to 1min of real world time. For instance, say the camera fps is running at 15fps and NOT 30fps, if I take 30*60 frames and encode the .avi at 30fps, the video will still have 1 min. However, it will correspond to 2min of real-time data. I suspect that OBS might be actually counting the frames and not real time. I find it unlikely that the camera will run at exactly the fps you set in OBS, but this is just a guess. One way to make sure is to have some sort of real-world alignment, say a timer, a record for a few minutes, and check if it matches the duration of the video.

**Reply by chongtianyifa** [reply]
I know the discrepancy concerning the duration is caused by the difference between the rate of camera capture and that of encoding by software. But we are talking about the relationship between the number of frames and the real recording duration.

1) Should be my bonsai test with timer your 'alternative way' as fps is defined at 30fps and duration at 2min? However, the number of frames is smaller than 30 x 60 x 2. 

2) Also, if 'your camera is likely acquiring frames at a slightly higher speed than 30fps', then more number of frames encoded at 30 fps will result in a video longer than 2 minutes rather shorter.

3) Importantly, my test is exactly what you said. The real duration is controlled by observing my timer of cellphone. The longer the real recording duration is, bonsai dropped more frames while OBS always gave a video with exact length and frame numbers.

**Reply by bruno-f-cruz** [reply]
Those 60x2 seconds, for bonsai, determine the duration the camera is acquired for, not the duration of the output encoded video. If you keep everything the same but change the encoding fps for 15fps, the encoded video will have double the duration. I guess what I am trying to say is that the duration of the output video is a bit meaningless for the most part. As [jbscherrer](https://github.com/jbscherrer) stated, what we usually care about is "when" does that data correspond to in "real-time" units.
Regardless, If you really want an output video of exactly 1min, saved at 30fps, use the Take operator instead of the timer, and set the Take property to fps*seconds. However, this approach rests on the assumption that your camera is acquiring at exactly the encoded fps of the output video which might not be the case as we discussed before.

PS: you are absolutely correct, I should have written less than 30fps, not more. Sorry!

**Reply by chongtianyifa** [reply]
Yes, I agree with that the duration of the output video is not important as it was just determined by the number of frames/the encoding fps defined at metadata. But I still don't get why OBS can save the exact number of frames matching the encoding fps*real time while Bonsai saved a smaller amount.

**Reply by bruno-f-cruz** [reply]
Not sure. My best guess is what I pointed out before. OBS tries to save a video with the duration you specified, even if that means recording more, or less, data from "real world time".

**Reply by aaronjnorris** [reply]
In my experience, I found "short" videos from Bonsai due to an average frame rate acquisition that is below the fps set in video writer. All of the captured frames are present - they are played back more quickly than they were recorded - in other words at a higher frame rate.  I wonder how standard video recording programs are able to correct for slower or changing incoming frame rates (ex shift to a darker scene triggering lower fps to lengthen exposure) to set the playback rate and if it is possible to do something similar in bonsai.

**Comment by IamJuanHilario** [comment]
@aaronjnorris Did you find a solution to this problem?, I am currently facing the same issue, but it extends to also not aquiring csv writer files to the specifed precision of for example  1 minute when I use a "timer" connected to a  "workflow out" module to end the experiment

**Reply by aaronjnorris** [reply]
No.  I continue to struggle with how Bonsai interacts with USB cameras.  I am sure I do not understand some fundamental concepts about how USB cameras are controlled. 
For some cameras, the amcap program here https://www.svpro.cc/download/ can help with setting settings that will be used by Bonsai despite not being accessible in menus in Bonsai. Making sure the exposure is set to manual is helpful in maintaining consistency. I don't know enough to understand how different programs access the USB camera settings or how the settings are negotiated between the camera and the program.  in some cases settings available amcap, for example, can not be set in bonsai.

For many parameters that we examine time between frames is important to know. I have taken to checking that the frame rate and interframe interval are what I think it is - often it is not what is set - I am guessing due to hardware limitations???
Here is a version of a script I use to add this information to the video directly and to check before recording what the  
[Video_FPSCheck.zip](https://github.com/bonsai-rx/bonsai/files/13442178/Video_FPSCheck.zip)
I try to make sure the actual frame rate, interframe interval, and the set encoding rate all jive.

If anyone has better options or insight into how to better control the frame rate, exposure, gain, etc of USB cameras from in Bonsai I would greatly appreciate the information.

**Reply by bruno-f-cruz** [reply]
I am not sure your question is related to this post. Can you open a new discussion post with the relevant example? Thanks.

Regarding the original post, I am still not sure what the actual question is. Bonsai will let you set a target fps exposure, etc... in the camera. Once that's done bonsai will just listen to incoming frames, it will not attempt to interfere with the data stream in any way. The same happens with videowriter, whatever frames the camera issues, the writer will encode and save to the file. It will not interpolate frames, it will not duplicate frames (something that obs does depending on your encoding settings). What you see is what you get (or the other way around in this case I guess haha)
